srcCpp/ParameterManager.cxx

```diff
@@ -0,0 +1,271 @@
+        peer->setDescription("Adds a peer to the peer host address list.\nThis argument may be repeated to indicate multiple peers");
+        peer->setType(T_VECTOR_STR);
+        peer->setExtraArgument(YES);
+        parameterList["peer"] = peer;
+    } catch(std::bad_alloc &ex) {
+        fprintf(stderr, "Exception in ParameterManager::initialize(): %s.\n", ex.what());
+        return false;
+    }
+
+    try {
+        std::vector<unsigned long long> _scanDataLenSizes;
+        _scanDataLenSizes.push_back(32);
+        _scanDataLenSizes.push_back(64);
+        _scanDataLenSizes.push_back(128);
+        _scanDataLenSizes.push_back(256);
+        _scanDataLenSizes.push_back(512);
+        _scanDataLenSizes.push_back(1024);
+        _scanDataLenSizes.push_back(2048);
+        _scanDataLenSizes.push_back(4096);
+        _scanDataLenSizes.push_back(8192);
+        _scanDataLenSizes.push_back(16384);
+        _scanDataLenSizes.push_back(32768);
+        _scanDataLenSizes.push_back(63000);
+        ParameterVector<unsigned long long> * scan = new ParameterVector<unsigned long long>(_scanDataLenSizes);
+        scan->setCommandLineArgument(std::make_pair("-scan","<size1>:<size2>:...:<sizeN>"));
+        scan->setDescription("Run test in scan mode, traversing\na range of sample data sizes from\n[32,63000] or [63001,2147483128] bytes,\nin the case that you are using large data or not.\nThe list of sizes is optional.\nDefault values are '32:64:128:256:512:1024:2048:4096:8192:16384:32768:63000'\nDefault: Not set\n");
+        scan->setType(T_VECTOR_NUMERIC);
+        scan->setExtraArgument(OPTIONAL);
+        scan->setRange(32, 2147483128);
+        scan->setParseMethod(SPLIT);
+        parameterList["scan"] = scan;
+    } catch(std::bad_alloc &ex) {
+        fprintf(stderr, "Exception in ParameterManager::initialize(): %s.\n", ex.what());
+        return false;
+    }
+
+    try {
+        Parameter<unsigned long long> *unbounded = new Parameter<unsigned long long>(63000);
+        unbounded->setCommandLineArgument(std::make_pair("-unbounded","<allocation_threshold>"));
+        unbounded->setDescription("Use unbounded Sequences\n<allocation_threshold> is optional, default 63000 Bytes.");
+        unbounded->setType(T_NUMERIC);
+        unbounded->setExtraArgument(OPTIONAL);
+        unbounded->setRange(0, 2147483128);
+        parameterList["unbounded"] = unbounded;
+    } catch(std::bad_alloc &ex) {
+        fprintf(stderr, "Exception in ParameterManager::initialize(): %s.\n", ex.what());
+        return false;
+    }
+
+    return true;
+}
+
+
+// Parse the command line parameters and set the value
+bool ParameterManager::parse(int argc, char *argv[])
+{
+    unsigned long long var;
+    bool success = true;
+    // Copy all arguments into a container of strings
+    std::vector<std::string> allArgs(argv, argv + argc);
+
+    std::map<std::string, Parameter_base *>::iterator it;
```

Comment:
Check out c++ type erasure
Look for dds/sub/AnyDataReader.hpp implementation

```

ParameterBase *param_base_;

template <ParameterType>
AnyParameter(ParameterType *param) : param_base(param){

}


template<ParamterType>
ParameterType* get()
{
   return static_cast<ParameterTpe>(param_);
}

template <tipename T>
ParameterVector<T>*get_vector;

std::map< strint, Anyparameter  > params;
AnyParameter param(new Parameter<loong>(...);
params["arg1] = ...;

Parameter<long> *param = params["argv1"].get<long> ()

AnyParameter param = params[argv1]; //BAD


   AnyParameter & operator=...(AnyParamer & other)
{
   other.param_ = param_
   param_ = NULL;
};
~

The class is not templated
Only the constructor and the get.
```

Reply:
Done